src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/LegacyCSProjPackageReferenceProject.cs

```diff
@@ -0,0 +1,341 @@
+/// </summary>
+    public class LegacyCSProjPackageReferenceProject : BuildIntegratedNuGetProject
+    {
+        private const string _assetsFileName = "project.assets.json";
+        private const string _includeAssets = "IncludeAssets";
+        private const string _excludeAssets = "ExcludeAssets";
+        private const string _privateAssets = "PrivateAssets";
+
+        private static Array _desiredPackageReferenceMetadata;
+
+        private readonly IEnvDTEProjectAdapter _project;
+
+        private IScriptExecutor _scriptExecutor;
+
+        static LegacyCSProjPackageReferenceProject()
+        {
+            _desiredPackageReferenceMetadata = Array.CreateInstance(typeof(string), 3);
+            _desiredPackageReferenceMetadata.SetValue(_includeAssets, 0);
+            _desiredPackageReferenceMetadata.SetValue(_excludeAssets, 1);
+            _desiredPackageReferenceMetadata.SetValue(_privateAssets, 2);
+        }
+
+        public LegacyCSProjPackageReferenceProject(
+            IEnvDTEProjectAdapter project)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            _project = project;
+
+            AssetsFilePath = Path.Combine(_project.GetBaseIntermediatePath().Result ?? "", _assetsFileName);
+            ProjectName = _project.Name;
+
+            InternalMetadata.Add(NuGetProjectMetadataKeys.Name, _project.Name);
+            InternalMetadata.Add(NuGetProjectMetadataKeys.UniqueName, _project.UniqueName);
+            InternalMetadata.Add(NuGetProjectMetadataKeys.FullPath, _project.ProjectFullPath);
+        }
+
+        #region IDependencyGraphProject
+
+        public override string AssetsFilePath { get; }
+
+        public override string ProjectName { get; }
+
+        /// <summary>
+        /// Making this timestamp as the current time means that a restore with this project in the graph
+        /// will never no-op. We do this to keep this work-around implementation simple.
+        /// </summary>
+        public override DateTimeOffset LastModified => DateTimeOffset.Now;
+
+        public override string MSBuildProjectPath => _project.ProjectFullPath;
+
+        private IScriptExecutor ScriptExecutor
+        {
+            get
+            {
+                if (_scriptExecutor == null)
+                {
+                    _scriptExecutor = ServiceLocator.GetInstanceSafe<IScriptExecutor>();
+                }
+
+                return _scriptExecutor;
+            }
+        }
+
+        public override async Task<IReadOnlyList<PackageSpec>> GetPackageSpecsAsync(DependencyGraphCacheContext context)
+        {
+            PackageSpec packageSpec;
+            if (context.PackageSpecCache.TryGetValue(_project.ProjectFullPath, out packageSpec))
+            {
+                return new[] { packageSpec };
+            }
+
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
GetPackageSpec has DTE pieces and they must be called on the UI thread or we get throws in DTE code.